### PR TITLE
Added BaseMappers and System Builder

### PR DESCRIPTION
**Pull request by KapilDuwadi**
- Added voltage, phase and equipment base mappers
- Added system_builder from distribution graph and mappers
- Updated docs

## Checklist

* [x] The pull request title is a good summary of the changes.
* [ ] Tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including "please review" to assign reviewers**

### CODE REVIEW

**Comment by KapilDuwadi** [comment]
I have not added tests for builder which requires create concrete classes for mappers. I will do that in next PR.

File: src/shift/graph/distribution_graph.py [hunk 1]
```diff
@@ -216,7 +233,7 @@ def get_node(self, node_name: str) -> NodeModel:
             msg = f"{node_name=} does not exist in the graph."
             raise NodeDoesNotExist(msg)
 
-        return self._graph.nodes[node_name]["node_data"]
+        return self._graph.nodes[node_name][self._node_data_ppty]
```
Comment (daniel-thom) [inline src/shift/graph/distribution_graph.py:236]: Line 232 is very slow if there are a lot of nodes. You create a list - O(n) - and then search that list - O(n).

File: src/shift/graph/distribution_graph.py [hunk 2]
```diff
@@ -239,7 +256,7 @@ def get_nodes(
         >>> dgraph.get_nodes()
         """
         for node in self._graph.nodes:
-            node_obj = self._graph.nodes[node]["node_data"]
+            node_obj = self._graph.nodes[node][self._node_data_ppty]
```
Comment (daniel-thom) [inline src/shift/graph/distribution_graph.py:259]: Does this need specific error handling for a bad input?

File: src/shift/system_builder.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        for node in self.dist_graph.get_nodes():
+            self._add_bus(node)
+            for asset in node.assets:
+                self._add_asset(node.name, asset)
+
+        for from_node, to_node, edge_data in self.dist_graph.get_edges():
+            equipment = self.equipment_mapper.edge_equipment_mapping[edge_data.name]
+            if type(equipment) != EQUIPMENT_TO_CLASS_TYPE.get(edge_data.edge_type):
+                msg = (
+                    f"{equipment=} is not supported for {edge_data=}"
```
Comment (daniel-thom) [inline src/shift/system_builder.py:92]: Nit: need extra space

File: src/shift/graph/distribution_graph.py [hunk 4]
```diff
@@ -50,8 +52,21 @@ class NodeModel(BaseModel):
 class EdgeModel(BaseModel):
     """Interface for edge model."""
 
+    model_config = ConfigDict(arbitrary_types_allowed=True)
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/graph/distribution_graph.py:55]: Sure we can create one base model.

File: src/shift/graph/distribution_graph.py [hunk 5]
```diff
@@ -2,8 +2,10 @@
 import copy
 
 import networkx as nx
-from pydantic import BaseModel, Field
-from shift.data_model import GeoLocation
+from pydantic import BaseModel, Field, model_validator, ConfigDict
+from gdm.quantities import PositiveDistance
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/graph/distribution_graph.py:6]: I think so. Let me double check.

File: src/shift/graph/distribution_graph.py [hunk 6]
```diff
@@ -50,8 +52,21 @@ class NodeModel(BaseModel):
 class EdgeModel(BaseModel):
     """Interface for edge model."""
 
+    model_config = ConfigDict(arbitrary_types_allowed=True)
     name: Annotated[str, Field(..., description="Name of the node.")]
     edge_type: Annotated[VALID_EDGE_TYPES, Field(..., description="Edge type.")]
+    length: Annotated[PositiveDistance | None, Field(None, description="Length of edge.")]
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/graph/distribution_graph.py:58]: Make sense.

File: src/shift/graph/distribution_graph.py [hunk 7]
```diff
@@ -50,8 +52,21 @@ class NodeModel(BaseModel):
...
+
+    @model_validator(mode="after")
+    def validate_fields(self):
+        if self.edge_type == DistributionTransformer and self.length is not None:
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/graph/distribution_graph.py:62]: Awh nice syntax.

File: src/shift/graph/distribution_graph.py [hunk 8]
```diff
@@ -50,8 +52,21 @@ class NodeModel(BaseModel):
...
+            msg = f"{self.length=} must be None for {self.edge_type=}"
+            raise ValueError(msg)
+
+        if self.edge_type == DistributionBranch and self.length is None:
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/graph/distribution_graph.py:66]: Awh I will revisit this. At this point I am okay with it.

File: src/shift/graph/distribution_graph.py [hunk 9]
```diff
@@ -100,6 +115,8 @@ class DistributionGraph:
     def __init__(self):
         self._graph = nx.Graph()
         self.vsource_node = None
+        self._node_data_ppty = "node_data"
+        self._edge_data_ppty = "edge_data"
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/graph/distribution_graph.py:119]: Yes. I need to move these.

File: src/shift/graph/openstreet_graph_builder.py [hunk 10]
```diff
@@ -235,22 +251,29 @@ def _get_distribution_graph_from_network(
 
         node_asset_mapper = self._get_node_assets_mapper(asset_nodes)
         for edge in graph.edges:
+            locs: list[Location] = []
             for node in edge:
                 if dist_graph.has_node(node):
                     continue
-                location = GeoLocation(graph.nodes[node]["x"], graph.nodes[node]["y"])
+                location = Location(
+                    x=graph.nodes[node]["x"], y=graph.nodes[node]["y"], crs="epsg:4326"
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/graph/openstreet_graph_builder.py:259]: I actually moved it in my next commit.

File: src/shift/graph/openstreet_graph_builder.py [hunk 11]
```diff
@@ -235,22 +251,29 @@ def _get_distribution_graph_from_network(
...
                 )
+                locs.append(location)
 
             dist_graph.add_edge(
                 *edge,
                 edge_data=EdgeModel(
                     name=str(uuid.uuid4()),
                     edge_type=DistributionBranch,
+                    length=get_distance_between_points(
+                        *[GeoLocation(loc.x, loc.y) for loc in locs]
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/graph/openstreet_graph_builder.py:275]: Geolocation is a named tuple.

File: src/shift/mapper/base_equipment_mapper.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    """Abstract class for mapping equipment to nodes and edges.
+
+    Subclasses must implement following method.
+    * transformer_voltage_mapping
+
+    Parameters
+    ----------
+
+    graph: DistributionGraph
+        Instance of `DistributionGraph` for which to implement voltage mapping.
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/mapper/base_equipment_mapper.py:19]: This I think only highlights it and does not add hyperlink.

File: src/shift/mapper/base_equipment_mapper.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    """
+
+    def __init__(self, graph: DistributionGraph):
+        self._graph = graph
+
+    @abstractmethod
+    @cached_property
+    def node_asset_equipment_mapping(self) -> dict[str, dict[VALID_NODE_TYPES, Component]]:
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/mapper/base_equipment_mapper.py:27]: Actually here it's not type We want full object.

File: src/shift/graph/distribution_graph.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/graph/distribution_graph.py:66]: I think this is necessary here we can discuss.

File: src/shift/system_builder.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+            Edge data object.
+        """
+        edge = edge_data.edge_type(
+            name=edge_data.name,
+            buses=[
+                self._system.get_component(DistributionBus, from_node),
+                self._system.get_component(DistributionBus, to_node),
+            ],
+            phases=self.phase_mapper.node_phase_mapping[from_node]
+            & self.phase_mapper.node_phase_mapping[to_node],
```
[the remark this answers is not in the document]
Comment (KapilDuwadi) [reply src/shift/system_builder.py:155]: It should be intersection of set of phases from both side.

**Pull request merged**